src/util/amount.rs

```diff
@@ -0,0 +1,362 @@
+    } else {
+        inner.checked_mul(10i64.pow(exp as u32)).unwrap()
+    }
+}
+
+fn pow_round_and_to_inner(v: f64, exp: i32) -> Inner {
+    let amt = v * 10f64.powi(exp);
+    if v < 0.0 {
+        (amt - 0.5) as Inner
+    } else {
+        (amt + 0.5) as Inner
+    }
+}
+
+#[derive(Clone, Copy)]
+pub struct Amount<P = Satoshi>(Inner, P)
+where
+    P: Precision;
+
+impl Amount<Satoshi> {
+    pub fn from_sat(sat: i64) -> Amount<Satoshi> {
+        Amount::from_inner(sat)
+    }
+}
+
+impl<P: Precision> Amount<P> {
+    pub fn from_inner(amount: Inner) -> Amount<P> {
+        Amount(amount, Default::default())
+    }
+
+    pub fn from_btc<T: IntoBtc<P>>(btc: T) -> Amount<P> {
+        btc.into_btc()
+    }
+
+    pub fn from_precision<O: Precision>(amount: Amount<O>) -> Amount<P> {
+        let amt = rescale_inner(amount.0, P::scale() - O::scale());
+        Amount::from_inner(amt)
+    }
+
+    pub fn as_precision<O: Precision>(&self) -> Amount<O> {
+        let amt = rescale_inner(self.0, O::scale() - P::scale());
+        Amount::from_inner(amt)
+    }
+
+    pub fn to_precision<O: Precision>(self) -> Amount<O> {
+        let amt = rescale_inner(self.0, O::scale() - P::scale());
+        Amount::from_inner(amt)
+    }
+
+    pub fn to_satoshi(self) -> Amount<Satoshi> {
+        self.to_precision::<Satoshi>()
+    }
+
+    pub fn as_satoshi(&self) -> Amount<Satoshi> {
+        self.as_precision::<Satoshi>()
+    }
+
+    pub fn zero() -> Amount<P> {
+        Amount::from_inner(0)
+    }
+
+    pub fn one() -> Amount<P> {
+        Amount::from_inner(1)
+    }
+
+    pub fn as_inner(&self) -> Inner {
+        self.0
+    }
+
+    pub fn into_inner(self) -> Inner {
+        self.0
+    }
+
+    pub fn float_precision<D: Precision>(&self) -> f64 {
+        let exp = D::scale() - P::scale();
+        (self.0 as f64) * 10f64.powi(exp)
+    }
+
+    pub fn fmt_denomination<D: Precision>(&self, f: &mut fmt::Write) -> fmt::Result {
+        write!(f, "{} {}", self.float_precision::<D>(), D::denom())
+    }
+
+    pub fn to_denominated_string<D: Precision>(&self) -> String {
+        let mut buf = String::new();
+        self.fmt_denomination::<D>(&mut buf).unwrap();
+        buf
+    }
+}
+
+impl<P: Precision> fmt::Debug for Amount<P> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}({})", P::denom(), self.0)
+    }
+}
+
+impl fmt::Display for Amount {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.fmt_denomination::<Bitcoin>(f)
+    }
+}
+
+impl<P: Precision> ops::Add for Amount<P> {
+    type Output = Amount<P>;
+
+    fn add(self, rhs: Amount<P>) -> Self::Output {
+        Amount::from_inner(self.0 + rhs.0)
+    }
+}
+
+impl<P: Precision> ops::Div for Amount<P> {
+    type Output = Amount<P>;
+
+    fn div(self, rhs: Amount<P>) -> Self::Output {
+        Amount::from_inner(self.0 / rhs.0)
+    }
+}
+
+impl<P: Precision> ops::Mul for Amount<P> {
```

Comment:
You will never want to multiply an `Amount` with another `Amount`. The unit of the result would be BTC^2 which for sure isn't represented by amount (and generally isn't really useful). Rhs should be u64 or something like that.

Reply:
Yeah I noticed this too. This is what rust-amount did. One problem I saw is that when you have multiplication with Number's, you also should have dividers, but dividing by an amount makes sense as well and AFAIK, having two dividers is not possible. Is it?

Reply:
I don't think so :/ but you could simply have two functions for that and no `Div` trait.